File: mobile-app/modules/sdk/android/src/main/java/expo/modules/glasssdk/glass/GlassEvent.kt
package expo.modules.glasssdk.glass

/**
 * Connection state enum for glass devices
 */
enum class GlassConnectionState {
    IDLE,           // No connection, not scanning
    SCANNING,       // Actively scanning for devices
    CONNECTING,     // Connection attempt in progress
    CONNECTED,      // Active connection established
    DISCONNECTING,  // Graceful disconnect in progress
    ERROR           // Error state
}

/**
 * Event constants for Glass module (70000 range to avoid conflicts)
 */
object GlassEventConstants {
    const val MSG_GLASS_DEVICE_FOUND = 70001
    const val MSG_GLASS_STATE_CHANGED = 70002
    const val MSG_GLASS_CONNECTED = 70003
    const val MSG_GLASS_DISCONNECTED = 70004
    const val MSG_GLASS_BATTERY = 70005
    const val MSG_GLASS_ERROR = 70006
    const val MSG_GLASS_STORAGE_INFO = 70007
    const val MSG_GLASS_SYSTEM_INFO = 70008
    const val MSG_GLASS_READY = 70009
    const val MSG_GLASS_SYNC_STARTED = 70010
    const val MSG_GLASS_SYNC_PROGRESS = 70011
    const val MSG_GLASS_SYNC_COMPLETED = 70012
    const val MSG_GLASS_WEAR_STATUS = 70013
    const val MSG_GLASS_VOICE_WAKEUP = 70014
    const val MSG_GLASS_BT_STATUS = 70015
    const val MSG_GLASS_STREAM_STATUS = 70016
    const val MSG_GLASS_VOICE_STOP = 70017
    const val MSG_GLASS_SYNC_FILE = 70018  // a single file finished syncing (s1=path, s2=type)
    const val MSG_GLASS_OTA_PROGRESS = 70019
    const val MSG_GLASS_OTA_COMPLETED = 70020
    const val MSG_GLASS_OTA_ERROR = 70021
    const val MSG_GLASS_OTA_LINUX_COMPLETED = 70022

    // Legacy K900 SDK events (from old EventMsg)
    const val MSG_QR_SCAN_RESULT = 10020
}

/**
 * Simple event message class for Glass module
 * Compatible with EventBus posting
 */
data class GlassEventMsg(
    val cmd: Int,
    val n1: Int = 0,
    val n2: Int = 0,
    val s1: String? = null,
    val s2: String? = null
)

/**
 * Event types for glass device communication
 */
object GlassEventType {
    const val DEVICE_FOUND = "GLASS_DEVICE_FOUND"
    const val STATE_CHANGED = "GLASS_STATE_CHANGED"
    const val CONNECTED = "GLASS_CONNECTED"
    const val DISCONNECTED = "GLASS_DISCONNECTED"
    const val BATTERY = "GLASS_BATTERY"
    const val ERROR = "GLASS_ERROR"
    const val VOICE_WAKEUP = "GLASS_VOICE_WAKEUP"
}

/**
 * Base class for all glass events
 */
sealed class GlassEvent(val type: String) {
    /**
     * Convert event to Map for React Native bridge
     */
    abstract fun toMap(): Map<String, Any>

    /**
     * Device discovered during scan
     */
    data class DeviceFound(
        val deviceId: String,
        val deviceName: String,
        val rssi: Int? = null
    ) : GlassEvent(GlassEventType.DEVICE_FOUND) {
        override fun toMap(): Map<String, Any> = buildMap {
            put("deviceId", deviceId)
            put("deviceName", deviceName)
            rssi?.let { put("rssi", it) }
        }
    }

    /**
     * Connection state changed
     */
    data class StateChanged(
        val oldState: GlassConnectionState,
        val newState: GlassConnectionState,
        val reason: String? = null
    ) : GlassEvent(GlassEventType.STATE_CHANGED) {
        override fun toMap(): Map<String, Any> = buildMap {
            put("oldState", oldState.name)
            put("newState", newState.name)
            reason?.let { put("reason", it) }
        }
    }

    /**
     * Successfully connected to device
     */
    data class Connected(
        val deviceId: String,
        val deviceName: String
    ) : GlassEvent(GlassEventType.CONNECTED) {
        override fun toMap(): Map<String, Any> = mapOf(
            "deviceId" to deviceId,
            "deviceName" to deviceName
        )
    }

    /**
     * Disconnected from device
     */
    data class Disconnected(
        val deviceId: String,
        val reason: String,
        val wasExpected: Boolean
    ) : GlassEvent(GlassEventType.DISCONNECTED) {
        override fun toMap(): Map<String, Any> = mapOf(
            "deviceId" to deviceId,
            "reason" to reason,
            "wasExpected" to wasExpected
        )
    }

    /**
     * Battery status update
     */
    data class Battery(
        val level: Int,
        val isCharging: Boolean,
        val voltage: Float? = null
    ) : GlassEvent(GlassEventType.BATTERY) {
        override fun toMap(): Map<String, Any> = buildMap {
            put("level", level)
            put("isCharging", isCharging)
            voltage?.let { put("voltage", it) }
        }
    }

    /**
     * Error occurred
     */
    data class Error(
        val error: String,
        val code: String? = null,
        val isFatal: Boolean = false
    ) : GlassEvent(GlassEventType.ERROR) {
        override fun toMap(): Map<String, Any> = buildMap {
            put("error", error)
            code?.let { put("code", it) }
            put("isFatal", isFatal)
        }
    }
}

/**
 * Storage information for the glass device
 */
data class GlassStorageInfo(
    val totalSize: Long,      // Total storage in bytes
    val freeSize: Long,       // Free storage in bytes
    val usedSize: Long,       // Used storage in bytes
    val photoCount: Int,      // Number of photos
    val videoCount: Int,      // Number of videos
    val audioCount: Int       // Number of audio files
)

/**
 * System information including battery status and device readiness
 */
data class GlassSystemInfo(
    val firmwareVersion: String,
    val appVersion: String,
    val deviceModel: String,
    val batteryLevel: Int,         // 0-100
    val isCharging: Boolean,
    val batteryVoltage: Float,    // Voltage in V
    val isReady: Boolean           // Device fully initialized and ready for use
)

File: mobile-app/modules/sdk/ios/glass/GlassEvent.swift
import Foundation

// MARK: - Connection State (mirrors GlassConnectionState Kotlin enum)
enum GlassConnectionState: String, CaseIterable {
    case idle          = "IDLE"
    case scanning      = "SCANNING"
    case connecting    = "CONNECTING"
    case connected     = "CONNECTED"
    case disconnecting = "DISCONNECTING"
    case error         = "ERROR"
}

// MARK: - Event Constants (mirrors GlassEventConstants Kotlin object)
struct GlassEventConstants {
    static let msgDeviceFound    = 70001
    static let msgStateChanged   = 70002
    static let msgConnected      = 70003
    static let msgDisconnected   = 70004
    static let msgBattery        = 70005
    static let msgError          = 70006
    static let msgStorageInfo    = 70007
    static let msgSystemInfo     = 70008
    static let msgReady          = 70009
    static let msgSyncStarted    = 70010
    static let msgSyncProgress   = 70011
    static let msgSyncCompleted  = 70012
    static let msgWearStatus     = 70013
    static let msgVoiceWakeup    = 70014
    static let msgQrScanResult   = 10020
    static let msgBtStatus       = 70015
    static let msgStreamStatus   = 70016
    static let msgVoiceStop       = 70017
    static let msgSyncFile       = 70018  // a single file finished syncing (s1=path, s2=type)
}

// MARK: - Internal notification name (replaces EventBus)
extension Notification.Name {
    static let glassEvent = Notification.Name("GlassSdkGlassEvent")
}

// MARK: - Event Message (mirrors GlassEventMsg Kotlin data class)
struct GlassEventMsg {
    let cmd: Int
    var n1: Int = 0
    var n2: Int = 0
    var s1: String? = nil
    var s2: String? = nil

    init(cmd: Int, n1: Int = 0, n2: Int = 0, s1: String? = nil, s2: String? = nil) {
        self.cmd = cmd
        self.n1 = n1
        self.n2 = n2
        self.s1 = s1
        self.s2 = s2
    }
}

// MARK: - Data structs (mirrors Kotlin data classes)

public struct GlassStorageInfo {
    var totalSize: Int64
    var freeSize: Int64
    var usedSize: Int64
    var photoCount: Int
    var videoCount: Int
    var audioCount: Int

    static let empty = GlassStorageInfo(
        totalSize: 0, freeSize: 0, usedSize: 0,
        photoCount: 0, videoCount: 0, audioCount: 0
    )

    func copy(
        totalSize: Int64? = nil, freeSize: Int64? = nil, usedSize: Int64? = nil,
        photoCount: Int? = nil, videoCount: Int? = nil, audioCount: Int? = nil
    ) -> GlassStorageInfo {
        GlassStorageInfo(
            totalSize: totalSize ?? self.totalSize,
            freeSize: freeSize ?? self.freeSize,
            usedSize: usedSize ?? self.usedSize,
            photoCount: photoCount ?? self.photoCount,
            videoCount: videoCount ?? self.videoCount,
            audioCount: audioCount ?? self.audioCount
        )
    }
}

public struct GlassSystemInfo {
    var firmwareVersion: String
    var appVersion: String
    var deviceModel: String
    var batteryLevel: Int
    var isCharging: Bool
    var batteryVoltage: Float
    var isReady: Bool

    static let empty = GlassSystemInfo(
        firmwareVersion: "", appVersion: "", deviceModel: "",
        batteryLevel: 0, isCharging: false, batteryVoltage: 0, isReady: false
    )

    func copy(
        firmwareVersion: String? = nil, appVersion: String? = nil,
        deviceModel: String? = nil, batteryLevel: Int? = nil,
        isCharging: Bool? = nil, batteryVoltage: Float? = nil, isReady: Bool? = nil
    ) -> GlassSystemInfo {
        GlassSystemInfo(
            firmwareVersion: firmwareVersion ?? self.firmwareVersion,
            appVersion: appVersion ?? self.appVersion,
            deviceModel: deviceModel ?? self.deviceModel,
            batteryLevel: batteryLevel ?? self.batteryLevel,
            isCharging: isCharging ?? self.isCharging,
            batteryVoltage: batteryVoltage ?? self.batteryVoltage,
            isReady: isReady ?? self.isReady
        )
    }
}

public struct PhotoResult {
    var success: Bool
    var errorCode: Int
    var photoCount: Int
    var filename: String?
}

public struct VideoResult {
    var success: Bool
    var errorCode: Int
    var videoCount: Int
}

public struct AudioResult {
    var success: Bool
    var errorCode: Int
    var audioCount: Int
}

public struct PhotoConfigData  { public var width: Int; public var height: Int }
public struct VideoConfigData  { public var width: Int; public var height: Int; public var quality: Int; public var duration: Int }
public struct AudioConfigData  { public var duration: Int }

public struct MediaConfigData {
    var photo: PhotoConfigData
    var video: VideoConfigData
    var audio: AudioConfigData
}
